File: 2019/6/solution.js
const fs = require("fs");
const path = require("path");

const input = fs
  .readFileSync(path.join(__dirname, "input.txt"), "utf-8", (err, data) => {
    if (err) {
      console.log(err);
    } else {
      return data;
    }
  })
  .split("\n");

const orbitalObjectMap = new Map();

class OrbitalObject {
  constructor(name) {
    this.name = name;
    this.orbiters = [];
    this.orbits = null;
    this.checked = false;
  }

  addOrbiter(orbitalObj) {
    this.orbiters.push(orbitalObj);
  }

  setOrbits(orbitalObj) {
    this.orbits = orbitalObj;
  }
}

for (const line of input) {
  const parts = line.split(")");
  if (!orbitalObjectMap.has(parts[0])) {
    orbitalObjectMap.set(parts[0], new OrbitalObject(parts[0]));
  }
  if (!orbitalObjectMap.has(parts[1])) {
    orbitalObjectMap.set(parts[1], new OrbitalObject(parts[1]));
  }
  const orbitee = orbitalObjectMap.get(parts[0]);
  const orbiter = orbitalObjectMap.get(parts[1]);
  orbitee.addOrbiter(orbiter);
  orbiter.setOrbits(orbitee);
}

function part1() {
  let res = 0;
  const queue = [...orbitalObjectMap.values()];
  while (queue.length) {
    const curr = queue.pop();
    res += curr.orbiters.length;
    queue.push(...curr.orbiters);
  }
  return res;
}

console.log(part1());

function part2(obj1, obj2) {
  const orbitalObj1 = orbitalObjectMap.get(obj1);
  orbitalObj1.checked = true;
  const queue = [[orbitalObj1, 0]];
  for (let i = 0; i < queue.length; i++) {
    const [currOrbitalObj, d] = queue[i];
    if (currOrbitalObj.name === obj2) {
      return d - 2;
    }
    currOrbitalObj.checked = true;
    if (currOrbitalObj.orbits && !currOrbitalObj.orbits.checked) {
      queue.push([currOrbitalObj.orbits, d + 1]);
    }
    for (const orbitalObj of currOrbitalObj.orbiters) {
      if (!orbitalObj.checked) {
        queue.push([orbitalObj, d + 1]);
      }
    }
  }
}

console.log(part2("YOU", "SAN"));

File: 2019/6/solution.go
package main

import (
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"strings"
)

func main() {
	_, filename, _, _ := runtime.Caller(0)
	dirname := filepath.Dir(filename)
	inputFilePath := filepath.Join(dirname, "input.txt")
	data, err := os.ReadFile(inputFilePath)
	if err != nil {
		panic(err)
	}
	puzzleInput := parseInput(strings.Split(string(data), "\n"))
	fmt.Println(part1(puzzleInput))
	fmt.Println(part2(puzzleInput))
}

type Orbiter struct {
	parent   *Orbiter
	children []*Orbiter
}

type QueueEntry struct {
	orbiter *Orbiter
	depth   int
}

func parseInput(data []string) map[string]*Orbiter {
	orbiterMap := make(map[string]*Orbiter)
	var name1, name2 string
	var obj1, obj2 *Orbiter
	for _, line := range data {
		objects := strings.Split(line, ")")
		name1 = objects[0]
		name2 = objects[1]
		if _, ok := orbiterMap[name1]; !ok {
			orbiterMap[name1] = &Orbiter{nil, []*Orbiter{}}
		}
		if _, ok := orbiterMap[name2]; !ok {
			orbiterMap[name2] = &Orbiter{nil, []*Orbiter{}}
		}
		obj1 = orbiterMap[name1]
		obj2 = orbiterMap[name2]
		obj1.children = append(obj1.children, obj2)
		obj2.parent = obj1
	}
	return orbiterMap
}

func part1(orbiterMap map[string]*Orbiter) int {
	res := 0
	var queue []QueueEntry
	queue = append(queue, QueueEntry{orbiterMap["COM"], 0})
	for len(queue) > 0 {
		entry := queue[0]
		queue = queue[1:]
		orbiter := entry.orbiter
		depth := entry.depth
		res += depth
		for _, child := range orbiter.children {
			queue = append(queue, QueueEntry{child, depth + 1})
		}
	}
	return res
}

func part2(orbiterMap map[string]*Orbiter) int {
	visited := make(map[*Orbiter]struct{})
	var queue []QueueEntry
	queue = append(queue, QueueEntry{orbiterMap["YOU"], 0})
	visited[orbiterMap["YOU"]] = struct{}{}
	for len(queue) > 0 {
		entry := queue[0]
		queue = queue[1:]
		orbiter := entry.orbiter
		depth := entry.depth
		if orbiter == orbiterMap["SAN"] {
			return depth - 2
		}
		for _, child := range orbiter.children {
			if _, ok := visited[child]; ok {
				continue
			}
			visited[child] = struct{}{}
			queue = append(queue, QueueEntry{child, depth + 1})
		}
		if orbiter.parent == nil {
			continue
		}
		if _, ok := visited[orbiter.parent]; !ok {
			visited[orbiter.parent] = struct{}{}
			queue = append(queue, QueueEntry{orbiter.parent, depth + 1})
		}
	}
	return -1
}
